### PR TITLE
[spi_device] 2nd part of the Technical Specification

### PR DESCRIPTION
**Pull request by eunchan**
This PR adds the remained flash mode and passthrough mode on top of #9951 .

### CODE REVIEW

File: hw/ip/spi_device/doc/_index.md [hunk 1]
```diff
@@ -309,6 +361,252 @@ All reads and writes to/from the SRAM for RXF and TXF activity are managed by
...
+
+### Command Information List
+
+The SW may configure the map from the received opcode to the command process module by programming *cmd_info* list.
+Current SPI_DEVICE provides 24 command information entries.
+Each entry represents a command.
+Details of the fields are explained in the {{<regref "CMD_INFO_0">}}
+
+First 11 commands are assigned to specific submodules.
+
```
Comment (kosta-kojdic) [inline hw/ip/spi_device/doc/_index.md:374]: Can you document required encodings for the first 11 commands? As of now all reset values are 0, so it is important to have proper commands configured in the first 11 slots.

File: hw/ip/spi_device/doc/_index.md [hunk 2]
```diff
@@ -19,6 +21,24 @@ title: "SPI Device HWIP Technical Specification"
 - Interrupts for RX/TX SRAM FIFO conditions (empty, full, designated level for
   RX, TX)
 
+### SPI Flash/ Passthrough Modes
+
+- Support Serial Flash emulation
+  - HW processed Read Status, Read JEDEC ID, Read SFDP, EN4B/ EX4B, and multiple read commands
+  - 16 depth Command/ Address FIFOs and Payload buffer for command upload
```
Comment (tjaychen) [inline hw/ip/spi_device/doc/_index.md:28]: could you mention the payload buffer size?

File: hw/ip/spi_device/doc/_index.md [hunk 3]
```diff
@@ -49,6 +69,38 @@ The SW can receive TPM commands with payload (address and data) and respond to t
...
 The submodule also supports the SW by managing a certain set of the FIFO registers and returning the read request by HW quickly.
 
+In Flash mode, SPI Device HWIP behaves as a Serial Flash device by recognizing SPI Flash commands and processing those commands by HW.
+The commands processed by HW are Read Status (1, 2, 3), Read JEDEC ID, Read SFDP, EN4B/ EX4B, and read commands with aids of SW.
+The IP supports Normal Read, Fast Read, Fast Read Dual Output, Fast Read Quad Output.
+This version of IP does not support Dual IO, Quad IO, QPI commands.
+
+In Passthrough mode, SPI Device receives SPI transactions from a host system and forwards the transactions to a downstream flash device.
+SW may filter prohibited commands by configuring 256-bit {{< regref "FILTER" >}} CSR.
+The IP cancels ongoing transaction if the received opcode matches to the filter CSR by de-asserting CSb and gating SCK  to the downstream flash device.
```
Comment (tjaychen) [inline hw/ip/spi_device/doc/_index.md:79]: extra space after SCK

File: hw/ip/spi_device/doc/_index.md [hunk 4]
```diff
@@ -49,6 +69,38 @@ The SW can receive TPM commands with payload (address and data) and respond to t
...
+
+SW may program CSRs to change the address and/or the first 4 bytes of payload on-the-fly in Passthrough mode.
+The address translation feature allows SW to maintain A/B binary images without aids of the host system.
+The payload translation may be used to change the payload of Write Status commands to not allow certain fiels to be modified.
```
Comment (tjaychen) [inline hw/ip/spi_device/doc/_index.md:83]: fiels -> fields

File: hw/ip/spi_device/doc/_index.md [hunk 5]
```diff
@@ -49,6 +69,38 @@ The SW can receive TPM commands with payload (address and data) and respond to t
...
+
+Flash mode and Passthrough mode may share the datapath.
```
Comment (tjaychen) [inline hw/ip/spi_device/doc/_index.md:85]: i think instead of saying "share the datapath", i feel like "active at the same time" might be clearer.

File: hw/ip/spi_device/doc/_index.md [hunk 6]
```diff
@@ -49,6 +69,38 @@ The SW can receive TPM commands with payload (address and data) and respond to t
...
+SW is recommended to filter the commands being processed internally.
+Mailbox is an exception as it shares the Read command opcode.
+
+### SPI Device Modes and Active Submodules
+
+SPI Device HWIP has three modes + TPM mode, which are "Generic" mode (also known as FwMode), "Flash" mode, and "Passthrough" mode.
+Generic mode is exclusive. Flash and Passthrough modes share many parts of the datapath.
+TPM mode only shares the SPI and has separate CSb port, which allows that the host sends TPM commands while other SPI mode is active.
+
+Mod      | FwMode | Status | JEDEC | SFDP | Mailbox | Read | Addr4B | Upload | Passthrough
```
Comment (tjaychen) [inline hw/ip/spi_device/doc/_index.md:96]: does the passthrough / passthru intercept box need a Y?

File: hw/ip/spi_device/doc/_index.md [hunk 7]
```diff
@@ -309,6 +361,252 @@ All reads and writes to/from the SRAM for RXF and TXF activity are managed by
...
+First 11 commands are assigned to specific submodules.
+
+Index  | Assigned Submodule
+-------|--------------------
+[2:0]  | Read Status
+[3]    | Read JEDEC ID
+[4]    | Read SFDP
+[10:5] | Read commands
+
+When Read Status, Read JEDEC ID commands are processed in Flash mode, the other fields in the command information entry are not used.
```
Comment (tjaychen) [inline hw/ip/spi_device/doc/_index.md:382]: just to make sure i understand, if we are in flash mode, OR we are in passthrough mode with intercept_en set, the other fields of these configurations are ignored right? If what i'm saying is true, it may be clearer to describe it that way.

File: hw/ip/spi_device/doc/_index.md [hunk 8]
```diff
@@ -309,6 +361,252 @@ All reads and writes to/from the SRAM for RXF and TXF activity are managed by
...
+![Command Parser block](cmdparse.svg)
+
+Command parser (*cmdparse*) processes the first byte of the SPI and activates the processing submodules depending on the received opcode and the *cmd_info* list described in the previous section.
+
+The cmdparse compares the recevied opcode with the *cmd_info.opcode* data structure.
+If any entry matches to the received opcode, the cmdparse hands over the matched command information entry with the index to the corresponding submodule.
+As explained in the [previous section](#command-information-list), the command parser checks the index to activate Read Status / Read JEDEC ID/ Read Command / Address 4B modules.
+Other than the first 11 slots and last two slots (the last two slots are not visible to SW), the cmdparse checks the *upload* field and activates the upload module if the field is set.
+
+SW may configure the submodules if the modules process the command in the passthrough mode or not by changing {{<regref "INTERCEPT_EN">}} CSR.
```
Comment (tjaychen) [inline hw/ip/spi_device/doc/_index.md:404]: just to double check, are you saying that the submodules will process the command in passthrough mode only if the `intercept_en` is set right?

File: hw/ip/spi_device/doc/_index.md [hunk 9]
```diff
@@ -309,6 +361,252 @@ All reads and writes to/from the SRAM for RXF and TXF activity are managed by
...
+
+### Status Control
+
+If the received command is one of the three read status commands, STATUS control module reigns SPI interface after the opcode.
```
Comment (tjaychen) [inline hw/ip/spi_device/doc/_index.md:408]: reigns? just to be sure, you mean the status control module will take over the spi interface right?

File: hw/ip/spi_device/doc/_index.md [hunk 10]
```diff
@@ -309,6 +361,252 @@ All reads and writes to/from the SRAM for RXF and TXF activity are managed by
...
+#### Address Handling
+
+For read commands such as Normal Read, Fast Read {Single/ Dual/ Quad} Output commands, the address comes through ID0 only.
+The state machine in this block shifts the address one-by-one and decrements the address counter register by 1.
+
+When it reaches the 4B address (`addr[2]`), the module triggers the DPSRAM state machine to fetch data from the DPSRAM.
+When the module receives `addr[0]`, at the positive edge of SCK, the module moves to appropriate command state based on the given CMD_INFO data.
+
+If the received address falls into mailbox address ragne and mailbox feature is enabled, the module turns on the mailbox selection bit.
+Then all out-going requests to the DPSRAM are forwarded to the mailbox section, not the read buffer section.
```
Comment (tjaychen) [inline hw/ip/spi_device/doc/_index.md:459]: just wanted to clarify one thing here. If the mailbox space is for example located ahead of other space, we DO NOT dynamically switch right? Ie, once you start in mailbox, even if you read into non-mailbox space, will we still remain in mailbox space? 

I'm not suggesting we change anything, just that it would be good to clarify the behavior.

File: hw/ip/spi_device/doc/_index.md [hunk 11]
```diff
@@ -309,6 +361,252 @@ All reads and writes to/from the SRAM for RXF and TXF activity are managed by
...
+
+The two CSRs omit unnecessary fields from the **CMD_INFO** data structure.
+The HW logic creates the default **CMD_INFO** structures for the two commands.
+The command parser module uses the generated structures to process and trigger the 4B management module.
+
+When the HW receives one of the commands, the HW changes the broadcast signal *cfg_addr_4b_en*.
+Also the HW updates {{<regref "CFG.addr_4b_en">}} after passing through CDC.
+It takes at most three SYS_CLK cycles to update the value in the *CFG* register after the completion of the SPI transaction (CSb de-assertion).
+
+_Note: The HW changes the broadcasting signal and the CSR even thought eh SPI host system sends more than 8 beats of the SPI S[0].
```
Comment (tjaychen) [inline hw/ip/spi_device/doc/_index.md:506]: nit, extra underscore in front of Note? or maybe that was intended.

Comment (tjaychen) [reply hw/ip/spi_device/doc/_index.md:506]: even thought -> even though

**Review by tjaychen — APPROVED** [review]
some nitpicks, but overall this looks great to me.

File: hw/ip/spi_device/doc/_index.md (hunk 3 above)
Comment (eunchan) [reply hw/ip/spi_device/doc/_index.md]: Blame my keyboard. It easily adds additional space ... I probably switch the switch .

File: hw/ip/spi_device/doc/_index.md (hunk 4 above)
Comment (eunchan) [reply hw/ip/spi_device/doc/_index.md:83]: Another blame to my kb 😀

File: hw/ip/spi_device/doc/_index.md (hunk 5 above)
Comment (eunchan) [reply hw/ip/spi_device/doc/_index.md:85]: I agree that is clearer. However, the mode register can't be active both. Flash mode or Passthrough.  In passthrough, flash mode can be active. That is intention. Let me rephrase the sentence.

File: hw/ip/spi_device/doc/_index.md (hunk 6 above)
Comment (eunchan) [reply hw/ip/spi_device/doc/_index.md]: Thanks! @msfschaffner also pointed out. Revised.

File: hw/ip/spi_device/doc/_index.md (hunk 7 above)
Comment (eunchan) [reply hw/ip/spi_device/doc/_index.md:382]: Right. Thanks for pointing it out.

File: hw/ip/spi_device/doc/_index.md (hunk 8 above)
Comment (eunchan) [reply hw/ip/spi_device/doc/_index.md]: Yes. Revised the sentence.

File: hw/ip/spi_device/doc/_index.md (hunk 9 above)
Comment (eunchan) [reply hw/ip/spi_device/doc/_index.md:407]: Yep. "take over" is way better 👍

File: hw/ip/spi_device/doc/_index.md (hunk 10 above)
Comment (eunchan) [reply hw/ip/spi_device/doc/_index.md:458]: Hum... good point. Let me check the design. I am honestly not sure about the behavior.

File: hw/ip/spi_device/doc/_index.md (hunk 1 above)
Comment (eunchan) [reply hw/ip/spi_device/doc/_index.md:374]: OK. Since I don't see any reason to have other than the fixed value for Read Status, Read JEDEC ID, and Read SFDP, I may follow the same process as `EN4B` / `EX4B`, which means, SW just configure *valid* and *opcode*.

I will create an issue and follow up.

File: hw/ip/spi_device/doc/_index.md [hunk 12]
```diff
@@ -309,6 +361,251 @@ All reads and writes to/from the SRAM for RXF and TXF activity are managed by
...
+
+The two CSRs omit unnecessary fields from the **CMD_INFO** data structure.
+The HW logic creates the default **CMD_INFO** structures for the two commands.
+The command parser module uses the generated structures to process and trigger the 4B management module.
+
+When the HW receives one of the commands, the HW changes the broadcast signal *cfg_addr_4b_en*.
+Also the HW updates {{<regref "CFG.addr_4b_en">}} after passing through CDC.
+It takes at most three SYS_CLK cycles to update the value in the *CFG* register after the completion of the SPI transaction (CSb de-assertion).
+
+_Note: The HW changes the broadcasting signal and the CSR even though eh SPI host system sends more than 8 beats of the SPI S[0].
```
Comment (msfschaffner) [inline hw/ip/spi_device/doc/_index.md:505]: nit: eh -> the

File: hw/ip/spi_device/doc/_index.md [hunk 13]
```diff
@@ -309,6 +361,251 @@ All reads and writes to/from the SRAM for RXF and TXF activity are managed by
...
+After the logic matches the received command byte with EN4B/ EX4B, the logic ignores the rest of the SPI data._
+
+The broadcasted `cfg_addr_4b_en` signal affects the read commands which `addr_mode` is *AddrCfg* in their command informatio entries.
```
Comment (msfschaffner) [inline hw/ip/spi_device/doc/_index.md:508]: nit: informatio -> information

File: hw/ip/spi_device/doc/_index.md [hunk 14]
```diff
@@ -309,6 +361,251 @@ All reads and writes to/from the SRAM for RXF and TXF activity are managed by
...
+
+### Command Upload
+
+If the received command meets following conditions, the HW stores the command into the command/ address FIFOs and the payload buffer:
+
+- The command does not match to the first 11 command information entries nor EN4B/ EX4B.
+- The command matches to any of the rest command information entries.
+- The matched entry has the `upload` field set.
+
+The upload module checks  the command information entry to determine whether the address/ payload fields to be uploaded or not.
```
Comment (msfschaffner) [inline hw/ip/spi_device/doc/_index.md:518]: nit: two spaces

File: hw/ip/spi_device/doc/_index.md [hunk 15]
```diff
@@ -309,6 +361,251 @@ All reads and writes to/from the SRAM for RXF and TXF activity are managed by
...
+SW must configure *payload_dir* to **PayloadIn** and *payload_en* to `4'b 0001` in order for the payload translation feature to work correctly.
+
+#### Output Enable Control
+
+Passthrough module controls the output enable signals on both host and downstream sides.
+Controlling the output enable ports is critical to not overdrive the PAD directions.
+The information of the pad enable and direction is given by SW.
+SW configures the address size, payload lanes, dummy size in **CMD_INFO** slots.
+
+If passthrough logic does not find valid commandinformation entry based on the received opcode, it assumes the command is **PayloadIn&** Single IO command.
```
Comment (msfschaffner) [inline hw/ip/spi_device/doc/_index.md:595]: nit: commandinformation -> command information

Comment (msfschaffner) [reply hw/ip/spi_device/doc/_index.md:595]: also, is the `&` intended in `**PayloadIn&**`?

**Review by msfschaffner — APPROVED** [review]
LGTM with a few nits

**Comment by eunchan** [comment]
I don't think this PR affects ROM binary and FPGA tests. Let me merge this.
image
